Atomic/sign.py

```diff
@@ -0,0 +1,89 @@
+from . import util
+from . import Atomic
+import os
+import tempfile
+from .atomic import AtomicError
+from .client import ids_same, add_shaw
+
+
+class Sign(Atomic):
+    def sign(self):
+        # TODO
+        # If an alternate sign key is provided, should we verify that it is valid?
+        # Should we check to see if there are valid GPG signatures?
+        # Atomic is run as sudo. Should we work around that?
+
+        if self.args.debug:
+            util.write_out(str(self.args))
+
+        # Verify all the images provided are valid
+        self.check_input_validity()
+
+        signature_path = util.get_signature_path(self.atomic_config)
+
+        signer = util.get_atomic_config_item(['default_signer']) if \
+            self.args.sign_key is None else self.args.sign_key
+
+        # Make sure the signature path exists
+        if not os.path.exists(signature_path):
+            raise ValueError("The path signature path {} "
+                             "does not exist".format(signature_path))
+
+        for sign_image in self.args.sign_targets:
+            local_inspect_info = self._inspect_image(image=sign_image)
+            sign_image_id = local_inspect_info['Id']
+            manifest = util.skopeo_inspect('docker://{}'.format(sign_image), args=['--raw'])
+
+            # Make sure the local image is the same as the image on the
+            # registry
+
+            if not ids_same(sign_image_id, manifest['config']['digest']):
+                raise ValueError("The local image {} differs from the "
+                                 "version on the registry".format(sign_image))
+            with tempfile.NamedTemporaryFile(mode="w", delete=False) as manifest_file:
+                manifest_file.write(str(manifest))
+                reg, image, tag = util.decompose(self.get_fq_name(local_inspect_info)) # pylint: disable=unused-variable
```

Comment:
This doesn’t work AFAICS; local inspect gives you _one_ of the local tags of that image, with no guarantee that it will be the right one if there are several.

AFAICS this should not be requiring a local image at all, and only use the identity from the command line = used by the repository.  Yes, it will need some canonicalization to the effect of https://github.com/docker/docker/blob/master/reference/reference.go#L192 (`normalize`) and https://github.com/docker/docker/blob/master/reference/reference.go#L140 (`WithDefaultTag`). I’m not sure whether there already is a Python library doing this.